runtime-modules/storage/src/lib.rs

```
@@ -1,20 +1,2373 @@
+                    &voucher_update,
+                    OperationType::Increase
+                );
+            }
+            if !remove_buckets.is_empty() {
+                BagManager::<T>::remove_storage_buckets(&bag_id, remove_buckets.clone());
+
+                Self::change_storage_buckets_vouchers(
+                    &remove_buckets,
+                    &voucher_update,
+                    OperationType::Decrease
+                );
+            }
+
+            Self::deposit_event(
+                RawEvent::StorageBucketsUpdatedForBag(bag_id, add_buckets, remove_buckets)
+            );
+        }
+
+        /// Cancel pending storage bucket invite. An invitation must be pending.
+        #[weight = 10_000_000] // TODO: adjust weight
+        pub fn cancel_storage_bucket_operator_invite(origin, storage_bucket_id: T::StorageBucketId){
+            T::ensure_working_group_leader_origin(origin)?;
+
+            let bucket = Self::ensure_storage_bucket_exists(&storage_bucket_id)?;
+
+            Self::ensure_bucket_pending_invitation_status(&bucket)?;
+
+            //
+            // == MUTATION SAFE ==
+            //
+
+            <StorageBucketById<T>>::mutate(storage_bucket_id, |bucket| {
+                bucket.operator_status = StorageBucketOperatorStatus::Missing;
+            });
+
+            Self::deposit_event(
+                RawEvent::StorageBucketInvitationCancelled(storage_bucket_id)
+            );
+        }
+
+        /// Invite storage bucket operator. Must be missing.
+        #[weight = 10_000_000] // TODO: adjust weight
+        pub fn invite_storage_bucket_operator(
+            origin,
+            storage_bucket_id: T::StorageBucketId,
+            operator_id: WorkerId<T>,
+        ){
+            T::ensure_working_group_leader_origin(origin)?;
+
+            let bucket = Self::ensure_storage_bucket_exists(&storage_bucket_id)?;
+
+            Self::ensure_bucket_missing_invitation_status(&bucket)?;
```

Should we _require_ that there is not even a pending invitation? Perhaps a bit overly strict for no clear benefit.

---

There are at least two possible states: 1. the same worker was invited, 2. another worker was invited.

Do we want to have the same outcome of the extrinsic with these states?

Should we silently replace the current invite (or maybe with a different event or even with a different event depending on the state)? 

Requiring to cancel the previous invite is a straightforward approach. However, this also increases the number of extrinsic calls (+1 for canceling). If you prefer to simplify usability - I recommend default invite replacement with the same event. Please, let me know if you still think we need to replace the invite.

---

> Please, let me know if you still think we need to replace the invite.

We are deleting the bucket here, so its not a matter of replacing, its just a matter of entirely dropping the bucket, which by implication drops storage state which reflects the invite.

---

Your comment relates to the `invite_storage_bucket_operator` extrinsic judging by line number 1357. But it looks you mean the `delete_storage_bucket` extrinsic which has the same `Self::ensure_bucket_missing_invitation_status(&bucket)?;` check. Yes, it is safe to delete the check here - will do.

---

Fixed in https://github.com/Joystream/joystream/pull/2407/commits/1ee491524e7a22cfbb8d2eac2fa4ecfce002b44e